File: packages/core-py/src/context_shunt/probe.py
"""Bounded filesystem probe used by the gate.

Scanning stops at 351 lines or the byte cap, whichever comes first, so probing a
multi-gigabyte file costs the same as probing a small one. A probe that stopped early is
``exact=False``, which the gate treats as unknown scale and blocks. Nothing is executed
and the file is never buffered.
"""

from __future__ import annotations

import os
import stat
from collections.abc import Iterator
from typing import BinaryIO

from .gate import ProbeResult, ProbeSelection
from .limits import DEFAULT_LIMITS, Limits
from .textindex import count_lines_bounded

_CHUNK = 16 * 1024


def _kind(mode: int) -> str:
    if stat.S_ISREG(mode):
        return "file"
    if stat.S_ISDIR(mode):
        return "directory"
    if stat.S_ISFIFO(mode):
        return "fifo"
    if stat.S_ISSOCK(mode):
        return "socket"
    if stat.S_ISBLK(mode) or stat.S_ISCHR(mode):
        return "device"
    return "other"


class FileProber:
    def __init__(self, limits: Limits = DEFAULT_LIMITS):
        self._limits = limits

    def __call__(self, path: str, selection: ProbeSelection | None = None) -> ProbeResult:
        selection = selection or ProbeSelection()
        try:
            st = os.lstat(path)
        except OSError:
            return ProbeResult(exists=False)
        if stat.S_ISLNK(st.st_mode):
            # A symlink is not a regular file for gate purposes; the path policy decides
            # whether the target may become a source at all.
            return ProbeResult(exists=True, kind="other")
        kind = _kind(st.st_mode)
        if kind != "file":
            return ProbeResult(exists=True, kind=kind)

        # Bind classification and sizing to one nonblocking, no-follow descriptor. This
        # closes the lstat/open swap window and prevents a swapped FIFO from hanging the
        # synchronous pre-tool hook.
        flags = (
            os.O_RDONLY
            | getattr(os, "O_CLOEXEC", 0)
            | getattr(os, "O_NOFOLLOW", 0)
            | getattr(os, "O_NONBLOCK", 0)
        )
        fd = -1
        try:
            fd = os.open(path, flags)
            before = os.fstat(fd)
            if not stat.S_ISREG(before.st_mode) or not _same_identity(before, st):
                return ProbeResult(exists=True, kind="other")
            with os.fdopen(fd, "rb", buffering=0) as handle:
                fd = -1
                if selection.mode == "metadata":
                    result = ProbeResult(exists=True, kind="file")
                elif selection.mode == "lines":
                    result = _selected_lines(
                        handle, selection.start_line, selection.limit, self._limits
                    )
                elif selection.mode == "tail":
                    result = _selected_tail(handle, before.st_size, selection.limit, self._limits)
                elif selection.mode == "search":
                    result = _search_upper_bound(handle, path, selection.max_matches, self._limits)
                else:
                    counted = count_lines_bounded(
                        _iter_handle(handle),
                        max_lines=self._limits.probe_max_lines_scanned,
                        # A full read is already blocked at this output byte cap.
                        max_bytes=self._limits.max_targeted_read_bytes + 1,
                    )
                    result = ProbeResult(
                        exists=True,
                        kind="file",
                        lines=counted.lines,
                        bytes=before.st_size if counted.exact else counted.bytes_seen,
                        exact=counted.exact,
                    )
                after = os.fstat(handle.fileno())
                return (
                    result
                    if _same_identity(before, after)
                    else ProbeResult(exists=True, kind="other")
                )
        except OSError:
            return ProbeResult(exists=True, kind="other")
        finally:
            if fd >= 0:
                os.close(fd)


def _same_identity(left: os.stat_result, right: os.stat_result) -> bool:
    return (
        left.st_dev,
        left.st_ino,
        left.st_size,
        left.st_mtime_ns,
        left.st_ctime_ns,
    ) == (
        right.st_dev,
        right.st_ino,
        right.st_size,
        right.st_mtime_ns,
        right.st_ctime_ns,
    )


def _iter_handle(handle: BinaryIO) -> Iterator[bytes]:
    while block := handle.read(_CHUNK):
        yield block


def _selected_lines(handle: BinaryIO, start_line: int, limit: int, limits: Limits) -> ProbeResult:
    end_line = start_line + limit
    line = 1
    lines = 0
    selected_bytes = 0
    scanned = 0
    current_selected = False
    while block := handle.read(_CHUNK):
        for byte in block:
            scanned += 1
            if scanned > limits.max_source_bytes:
                return ProbeResult(exists=True, lines=lines, bytes=selected_bytes, exact=False)
            current_selected = start_line <= line < end_line
            if current_selected:
                selected_bytes += 1
                if selected_bytes > limits.max_targeted_read_bytes:
                    return ProbeResult(exists=True, lines=lines, bytes=selected_bytes, exact=True)
            if byte == 0x0A:
                if current_selected:
                    lines += 1
                line += 1
                if line >= end_line:
                    return ProbeResult(exists=True, lines=lines, bytes=selected_bytes, exact=True)
    if current_selected and selected_bytes:
        lines += 1
    return ProbeResult(exists=True, lines=lines, bytes=selected_bytes, exact=True)


def _selected_tail(handle: BinaryIO, file_bytes: int, limit: int, limits: Limits) -> ProbeResult:
    if file_bytes == 0:
        return ProbeResult(exists=True)
    trailing_newline = _read_exact_at(handle, file_bytes - 1, 1) == b"\n"
    needed_newlines = limit + (1 if trailing_newline else 0)
    found = 0
    cursor = file_bytes
    while cursor > 0:
        start = max(0, cursor - _CHUNK)
        block = _read_exact_at(handle, start, cursor - start)
        for index in range(len(block) - 1, -1, -1):
            if block[index] == 0x0A:
                found += 1
                if found == needed_newlines:
                    selected_bytes = file_bytes - (start + index + 1)
                    return ProbeResult(exists=True, lines=limit, bytes=selected_bytes, exact=True)
        cursor = start
        if file_bytes - cursor > limits.max_targeted_read_bytes:
            return ProbeResult(
                exists=True,
                lines=limit,
                bytes=limits.max_targeted_read_bytes + 1,
                exact=True,
            )
    return ProbeResult(exists=True, lines=limit, bytes=file_bytes, exact=True)


def _read_exact_at(handle: BinaryIO, position: int, length: int) -> bytes:
    handle.seek(position)
    out = bytearray()
    while len(out) < length:
        block = handle.read(length - len(out))
        if not block:
            raise OSError("short read")
        out.extend(block)
    return bytes(out)


def _search_upper_bound(
    handle: BinaryIO, path: str, max_matches: int, limits: Limits
) -> ProbeResult:
    prefix_bytes = len(path.encode("utf-8")) + 64
    bytes_seen = 0
    line_bytes = 0
    line_count = 0
    max_line_bytes = 0
    ended_with_newline = False
    while block := handle.read(_CHUNK):
        for byte in block:
            bytes_seen += 1
            if bytes_seen > limits.max_source_bytes:
                return ProbeResult(exists=True, lines=line_count, bytes=bytes_seen, exact=False)
            line_bytes += 1
            current_upper = max_matches * (line_bytes + prefix_bytes)
            if current_upper > limits.max_targeted_read_bytes:
                return ProbeResult(
                    exists=True,
                    lines=min(max_matches, line_count + 1),
                    bytes=current_upper,
                    exact=True,
                )
            ended_with_newline = byte == 0x0A
            if ended_with_newline:
                line_count += 1
                max_line_bytes = max(max_line_bytes, line_bytes)
                line_bytes = 0
                upper = max_matches * (max_line_bytes + prefix_bytes)
                if upper > limits.max_targeted_read_bytes:
                    return ProbeResult(exists=True, lines=max_matches, bytes=upper, exact=True)
    if line_bytes or (bytes_seen and not ended_with_newline):
        line_count += 1
        max_line_bytes = max(max_line_bytes, line_bytes)
    matches = min(max_matches, line_count)
    return ProbeResult(
        exists=True,
        lines=matches,
        bytes=matches * (max_line_bytes + prefix_bytes),
        exact=True,
    )

File: packages/core-ts/src/probe.ts
/**
 * Bounded filesystem probe used by the gate.
 *
 * Scanning stops at 351 lines or the byte cap, whichever comes first, so probing a
 * multi-gigabyte file costs the same as probing a small one. A probe that stopped early is
 * `exact: false`, which the gate treats as unknown scale and blocks.
 */
import {
  type BigIntStats,
  closeSync,
  constants,
  fstatSync,
  lstatSync,
  openSync,
  readSync,
} from "node:fs";

import { ProbeResult, ProbeSelection } from "./gate.js";
import { DEFAULT_LIMITS, Limits } from "./limits.js";
import { countLinesBounded } from "./textindex.js";

const CHUNK = 16 * 1024;

function* readBlocks(fd: number): Generator<Uint8Array> {
  for (;;) {
    const buf = Buffer.allocUnsafe(CHUNK);
    const read = readSync(fd, buf, 0, CHUNK, null);
    if (read === 0) return;
    yield buf.subarray(0, read);
  }
}

function kindOf(st: BigIntStats): NonNullable<ProbeResult["kind"]> {
  if (st.isDirectory()) return "directory";
  if (st.isFIFO()) return "fifo";
  if (st.isSocket()) return "socket";
  if (st.isBlockDevice() || st.isCharacterDevice()) return "device";
  return st.isFile() ? "file" : "other";
}

function sameIdentity(left: BigIntStats, right: BigIntStats): boolean {
  return left.dev === right.dev
    && left.ino === right.ino
    && left.size === right.size
    && left.mtimeNs === right.mtimeNs
    && left.ctimeNs === right.ctimeNs;
}

export function fileProber(limits: Limits = DEFAULT_LIMITS) {
  return (path: string, selection: ProbeSelection = { mode: "full" }): ProbeResult => {
    let lst: BigIntStats;
    try {
      lst = lstatSync(path, { bigint: true });
    } catch {
      return { exists: false };
    }
    if (lst.isSymbolicLink()) {
      // A symlink is not a regular file for gate purposes; the path policy decides
      // whether the target may become a source at all.
      return { exists: true, kind: "other" };
    }
    const initialKind = kindOf(lst);
    if (initialKind !== "file") return { exists: true, kind: initialKind };

    // Bind classification and sizing to one nonblocking, no-follow descriptor. Without
    // this, an attacker could replace the lstat'd path with a symlink or FIFO before the
    // second open, escaping the path decision or hanging the synchronous host hook.
    let fd: number;
    try {
      fd = openSync(path, constants.O_RDONLY | constants.O_NOFOLLOW | constants.O_NONBLOCK);
    } catch {
      return { exists: true, kind: "other" };
    }
    try {
      const before = fstatSync(fd, { bigint: true });
      if (!before.isFile() || !sameIdentity(before, lst)) {
        return { exists: true, kind: "other" };
      }
      let result: ProbeResult;
      if (selection.mode === "metadata") {
        result = { exists: true, kind: "file", lines: 0, bytes: 0, exact: true };
      } else if (selection.mode === "lines") {
        result = selectedLines(fd, selection.startLine, selection.limit, limits);
      } else if (selection.mode === "tail") {
        result = selectedTail(fd, Number(before.size), selection.limit, limits);
      } else if (selection.mode === "search") {
        result = searchUpperBound(fd, path, selection.maxMatches, limits);
      } else {
        const counted = countLinesBounded(readBlocks(fd), {
          maxLines: limits.probeMaxLinesScanned,
          // A full read is blocked at the targeted-output byte cap, so reading beyond
          // the first byte over that cap cannot change the decision.
          maxBytes: limits.maxTargetedReadBytes + 1,
        });
        result = {
          exists: true,
          kind: "file",
          lines: counted.lines,
          bytes: counted.exact ? Number(before.size) : counted.bytesSeen,
          exact: counted.exact,
        };
      }
      const after = fstatSync(fd, { bigint: true });
      return sameIdentity(before, after) ? result : { exists: true, kind: "other" };
    } catch {
      return { exists: true, kind: "other" };
    } finally {
      closeSync(fd);
    }
  };
}

function selectedLines(fd: number, startLine: number, limit: number, limits: Limits): ProbeResult {
  const endLine = startLine + limit;
  let line = 1;
  let lines = 0;
  let bytes = 0;
  let scanned = 0;
  let currentSelected = false;
  for (;;) {
    const buffer = Buffer.allocUnsafe(CHUNK);
    const count = readSync(fd, buffer, 0, buffer.length, null);
    if (count === 0) break;
    for (let index = 0; index < count; index += 1) {
      scanned += 1;
      if (scanned > limits.maxSourceBytes) {
        return { exists: true, kind: "file", lines, bytes, exact: false };
      }
      currentSelected = line >= startLine && line < endLine;
      if (currentSelected) {
        bytes += 1;
        if (bytes > limits.maxTargetedReadBytes) {
          return { exists: true, kind: "file", lines, bytes, exact: true };
        }
      }
      if (buffer[index] === 0x0a) {
        if (currentSelected) lines += 1;
        line += 1;
        if (line >= endLine) {
          return { exists: true, kind: "file", lines, bytes, exact: true };
        }
      }
    }
  }
  if (currentSelected && bytes > 0) lines += 1;
  return { exists: true, kind: "file", lines, bytes, exact: true };
}

function selectedTail(
  fd: number,
  fileBytes: number,
  limit: number,
  limits: Limits,
): ProbeResult {
  if (fileBytes === 0) return { exists: true, kind: "file", lines: 0, bytes: 0, exact: true };
  const last = Buffer.allocUnsafe(1);
  readExactAt(fd, last, fileBytes - 1);
  const neededNewlines = limit + (last[0] === 0x0a ? 1 : 0);
  let found = 0;
  let cursor = fileBytes;
  while (cursor > 0) {
    const start = Math.max(0, cursor - CHUNK);
    const length = cursor - start;
    const buffer = Buffer.allocUnsafe(length);
    readExactAt(fd, buffer, start);
    for (let index = length - 1; index >= 0; index -= 1) {
      if (buffer[index] === 0x0a) {
        found += 1;
        if (found === neededNewlines) {
          const bytes = fileBytes - (start + index + 1);
          return { exists: true, kind: "file", lines: limit, bytes, exact: true };
        }
      }
    }
    cursor = start;
    const selectedSoFar = fileBytes - cursor;
    if (selectedSoFar > limits.maxTargetedReadBytes) {
      return {
        exists: true,
        kind: "file",
        lines: limit,
        bytes: limits.maxTargetedReadBytes + 1,
        exact: true,
      };
    }
  }
  return { exists: true, kind: "file", lines: limit, bytes: fileBytes, exact: true };
}

function readExactAt(fd: number, buffer: Buffer, position: number): void {
  let offset = 0;
  while (offset < buffer.length) {
    const count = readSync(fd, buffer, offset, buffer.length - offset, position + offset);
    if (count === 0) throw new Error("short read");
    offset += count;
  }
}

function searchUpperBound(fd: number, path: string, maxMatches: number, limits: Limits): ProbeResult {
  // Conservative allowance for path, line number, separators and host result framing.
  const prefixBytes = new TextEncoder().encode(path).length + 64;
  let bytesSeen = 0;
  let lineBytes = 0;
  let lineCount = 0;
  let maxLineBytes = 0;
  let endedWithNewline = false;
  for (;;) {
    const buffer = Buffer.allocUnsafe(CHUNK);
    const count = readSync(fd, buffer, 0, buffer.length, null);
    if (count === 0) break;
    for (let index = 0; index < count; index += 1) {
      bytesSeen += 1;
      if (bytesSeen > limits.maxSourceBytes) {
        return { exists: true, kind: "file", lines: lineCount, bytes: bytesSeen, exact: false };
      }
      lineBytes += 1;
      const currentUpper = maxMatches * (lineBytes + prefixBytes);
      if (currentUpper > limits.maxTargetedReadBytes) {
        return {
          exists: true,
          kind: "file",
          lines: Math.min(maxMatches, lineCount + 1),
          bytes: currentUpper,
          exact: true,
        };
      }
      endedWithNewline = buffer[index] === 0x0a;
      if (endedWithNewline) {
        lineCount += 1;
        maxLineBytes = Math.max(maxLineBytes, lineBytes);
        lineBytes = 0;
        const upper = maxMatches * (maxLineBytes + prefixBytes);
        if (upper > limits.maxTargetedReadBytes) {
          return { exists: true, kind: "file", lines: maxMatches, bytes: upper, exact: true };
        }
      }
    }
  }
  if (lineBytes > 0 || (bytesSeen > 0 && !endedWithNewline)) {
    lineCount += 1;
    maxLineBytes = Math.max(maxLineBytes, lineBytes);
  }
  const matches = Math.min(maxMatches, lineCount);
  return {
    exists: true,
    kind: "file",
    lines: matches,
    bytes: matches * (maxLineBytes + prefixBytes),
    exact: true,
  };
}
